### ISSUE TITLE
Respect locale settings

### ISSUE THREAD

**Issue by yooakim** [malxau/yori#18]
First off, thanks for taking the time to write Yori! Much appreciated :-)

I'm swedish and I usually run my operating system in English but with the Swedish locale. If I do a dir listing in classic cmd it looks as this:

```
 Volume in drive C is OS
 Volume Serial Number is F614-9BE2

 Directory of C:\

2019-11-05  12:52    <DIR>          dell
2019-11-05  10:37    <DIR>          Intel
2019-11-17  17:32    <DIR>          Microsoft
2019-11-11  13:36    <DIR>          NHibernate Mapping File Generator
2019-03-19  05:52    <DIR>          PerfLogs
2020-01-13  09:57    <DIR>          Program Files
2019-12-10  09:06    <DIR>          Program Files (x86)
2020-01-07  08:48    <DIR>          Temp
2019-12-02  08:40    <DIR>          Users
2019-12-10  09:15    <DIR>          Util
2020-01-08  14:48    <DIR>          Windows
2019-11-11  13:56    <DIR>          XactTestEnvironment
              12 Dir(s)  338 761 736 192 bytes free
```
if I do the same dir listing using Yori! I get the following output:

```

 Directory of C:\

2019/12/02  08:40  <DIR>             $Recycle.Bin
2019/11/05  12:52  <DIR>             dell
2019/05/24  18:51  <JUNCTION>        Documents and Settings [C:\Users]
2019/11/05  10:37  <DIR>             Intel
2019/11/05  19:31  <DIR>             IntelOptaneData
2019/11/17  17:32  <DIR>             Microsoft
2019/11/11  13:36  <DIR>             NHibernate Mapping File Generator
2020/01/13  08:30  <DIR>             OneDriveTemp
2019/03/19  05:52  <DIR>             PerfLogs
2020/01/13  09:57  <DIR>             Program Files
2019/12/10  09:06  <DIR>             Program Files (x86)
2020/01/13  09:25  <DIR>             ProgramData
2019/05/24  19:46  <DIR>             Recovery
2020/01/12  14:28  <DIR>             System Volume Information
2020/01/07  08:48  <DIR>             Temp
2019/12/02  08:40  <DIR>             Users
2019/12/10  09:15  <DIR>             Util
2020/01/08  14:48  <DIR>             Windows
2019/11/11  13:56  <DIR>             XactTestEnvironment
          18 Dir(s)     338,761,232,384 bytes free
```

As shown the date formatting is not using the Swedish locale format.

Not a big issue but it would be very nice if Yori used the same format as the rest of Windows is using for dates, times and numbers.

**Comment by malxau** [comment]
I dug into this a little, but think I should set expectations: CMD isn't doing what the rest of Windows is doing.  It's always including leading zeroes to help alignment, even though leading zeroes are part of the locale settings and are generally not used elsewhere.  This means CMD is left trying to reconstruct strings based on individual locale primitives (see GetLocaleInfo) rather than just let the system format things (as in GetDateFormat).

I'm heading down the same path, but because it's different and needs to be aware of every locale setting to be accurate, it will likely always be a little bit quirky.

**Comment by malxau** [comment]
I pushed a change to master to use locale aware settings for time and date.  As noted above, this isn't ideal: it currently covers five elements:

- Date separator character
- Time separator character
- Date field order
- Two or four character years
- 24 hour or 12 hour time

Looking at the other tools, sdir might want to adopt the first two or three, but since its goal is to be compact I don't think 12 hour time would work well there.  Finfo is really leaving the user in control of specifying the format string.  I can't think of anything else that deals in times and dates.

**Comment by aleaksunder** [comment]
Just want to know... Can this locale-awareness may break Yori's scripts universality?
We know that cmd.exe locales broke so many opportunities to write really universal scripts which would work on every system and every locale! For example `date` and `time` commands outputs in user-locale format, so we don't have opportunity to grab seperate year from that output and be sure that our code will work on every system and locale

You could just carefully look at theese links:
[https://www.robvanderwoude.com/datetimentparse.php](url)
[https://ritchielawrence.github.io/batchfunctionlibrary](url)
[https://ss64.com/nt/syntax-getdate.html](url)

I just want to point that the cmd's locale awareness just put all of scipters in situation where this is completely impossible to write script that get such a little thing as get current date in bullet-proof manner... This was discusting to not be able to do so!

AFAIK now Yori support output in English locale formats and this makes it predictable and reliable... Please do not set locale dependency by default... because we have some fantastic languages like Turkish, Chineese and Japanese where things go completely wrong way as in English one!

Or at least make an option for locale-awareness like command line parameter or maybe ini file parameter

**Comment by malxau** [comment]
That's a very valid concern and I had the same concern.  Just to note a few things:

- Commands line `date` and `finfo` output data points in a caller-specified format.  This was an attempt to make life better for scripts; rather than output text and have a script try to parse it, it allows the output to be in the form the script wants.  Those commands are not changing.
- Following on from the above, commands like `dir` or `sdir` seem like horrible things for scripts to parse, so my focus is on ensuring other tools are available for scripts.  
- The behavior here is optional and can be disabled with `-g`.

I totally agree with your high order point that, one way or another, there should be output that is easy for scripts to consume.

**Comment by malxau** [comment]
I just released 1.21 which I believe fixes this issue.  You can upgrade an existing install with `ypm -u`.  If you find any additional problems, please either reopen this issue or file a new one describing that problem.  Thanks for the report!
